contracts/simple_dex/lib.rs

```diff
@@ -0,0 +1,659 @@
+
+#[ink::contract]
+mod simple_dex {
+
+    use access_control::{traits::AccessControlled, Role, ACCESS_CONTROL_PUBKEY};
+    use game_token::{
+        ALLOWANCE_SELECTOR, BALANCE_OF_SELECTOR, TRANSFER_FROM_SELECTOR, TRANSFER_SELECTOR,
+    };
+    use ink_env::{
+        call::{build_call, Call, ExecutionInput, Selector},
+        CallFlags, DefaultEnvironment, Error as InkEnvError,
+    };
+    use ink_lang::{
+        codegen::{initialize_contract, EmitEvent},
+        reflect::ContractEventBase,
+    };
+    use ink_prelude::{format, string::String, vec};
+    use ink_storage::{traits::SpreadAllocate, Mapping};
+    use num::integer::Roots;
+    use openbrush::contracts::traits::errors::PSP22Error;
+
+    type Event = <SimpleDex as ContractEventBase>::Type;
+
+    #[derive(Debug, PartialEq, Eq, scale::Encode, scale::Decode)]
+    #[cfg_attr(feature = "std", derive(scale_info::TypeInfo))]
+    pub enum DexError {
+        PSP22(PSP22Error),
+        InsufficientAllowanceOf(AccountId),
+        Arithmethic,
+        MissingRole(Role),
+        InkEnv(String),
+        CrossContractCall(String),
+        NativeTransferFailed(String),
+    }
+
+    impl From<PSP22Error> for DexError {
+        fn from(e: PSP22Error) -> Self {
+            DexError::PSP22(e)
+        }
+    }
+
+    impl From<InkEnvError> for DexError {
+        fn from(why: InkEnvError) -> Self {
+            DexError::InkEnv(format!("{:?}", why))
+        }
+    }
+
+    #[ink(event)]
+    #[derive(Debug)]
+    pub struct NativeToToken {
+        caller: AccountId,
+        #[ink(topic)]
+        token_out: AccountId,
+        amount_in: Balance,
+        amount_out: Balance,
+    }
+
+    #[ink(event)]
+    #[derive(Debug)]
+    pub struct TokenToNative {
+        caller: AccountId,
+        #[ink(topic)]
+        token_in: AccountId,
+        amount_in: Balance,
+        amount_out: Balance,
+    }
+
+    #[ink(event)]
+    #[derive(Debug)]
+    pub struct TokenToToken {
+        caller: AccountId,
+        #[ink(topic)]
+        token_in: AccountId,
+        #[ink(topic)]
+        token_out: AccountId,
+        amount_in: Balance,
+        amount_out: Balance,
+    }
+
+    #[ink(event)]
+    #[derive(Debug)]
+    pub struct NativeDeposit {
+        caller: AccountId,
+        amount: Balance,
+        minted_liquidity: Balance,
+    }
+
+    #[ink(event)]
+    #[derive(Debug)]
+    pub struct NativeWithdrawal {
+        caller: AccountId,
+        amount: Balance,
+        redeemed_liquidity: Balance,
+    }
+
+    #[ink(event)]
+    #[derive(Debug)]
+    pub struct SwapFeeSet {
+        #[ink(topic)]
+        caller: AccountId,
+        swap_fee: Balance,
+    }
+
+    #[ink(storage)]
+    #[derive(SpreadAllocate)]
+    pub struct SimpleDex {
+        /// total supply of pool shares
+        pub total_liquidity: u128,
+        /// tracks pool shares per account
+        pub liquidity: Mapping<AccountId, u128>,
+        pub swap_fee: u128,
```

Comment:
What is it being set to by default? I don't see it being set in the constructor of the contract and if no one calls `set_swap_fee` before the first deal is made then it will use the default?